### PR TITLE
Add global state management

### PR DESCRIPTION
**Pull request by bsneed**
- Brings a global state object for important bits, such as user info and context data.
- The funnel now starts with a fully filled in context payload.
- Device-mode destinations can now utilize the context data that was previously only being sent to segment.com

### CODE REVIEW

File: Analytics/Classes/SEGAnalytics.h [hunk 1]
```diff
@@ -218,7 +218,11 @@ NS_SWIFT_NAME(Analytics)
 /** Returns the configuration used to create the analytics client. */
 - (SEGAnalyticsConfiguration *)configuration;
 
-
+/*
+ // a better way?
```
[the remark this answers is not in the document]
Comment (bsneed) [reply Analytics/Classes/SEGAnalytics.h:222]: I should just remove the comment.  I didn't end up making these change anyway.

File: Analytics/Classes/SEGSegmentIntegration.m [hunk 2]
```diff
@@ -463,27 +309,20 @@ - (void)enqueueAction:(NSString *)action dictionary:(NSMutableDictionary *)paylo
 {
     // attach these parts of the payload outside since they are all synchronous
     payload[@"type"] = action;
-    payload[@"messageId"] = GenerateUUIDString();
+    //payload[@"messageId"] = GenerateUUIDString();
```
Comment (migs647) [inline Analytics/Classes/SEGSegmentIntegration.m:312]: Are we keeping this?

Comment (bsneed) [reply Analytics/Classes/SEGSegmentIntegration.m:312]: No.  Will fix.

File: Analytics/Internal/SEGState.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+//
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+@class SEGAnalyticsConfiguration;
+
+@interface SEGUserInfo: NSObject
+@property (nonatomic, strong, nonnull) NSString *anonymousId;
```
Comment (migs647) [inline Analytics/Internal/SEGState.h:16]: No need to mark nonnull when using NS_ASSUME_NONNULL

File: Analytics/Internal/SEGState.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+@property (nonatomic, strong, nullable) NSString *userId;
+@property (nonatomic, strong, nullable) NSDictionary *traits;
+@end
+
+@interface SEGPayloadContext: NSObject
+@property (nonatomic, readonly, nonnull) NSDictionary *payload;
```
Comment (migs647) [inline Analytics/Internal/SEGState.h:22]: ^^

File: Analytics/Internal/SEGState.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+- (void)updateStaticContext;
+
+@end
+
+
+
+@interface SEGState : NSObject
+
+@property (nonatomic, readonly, nonnull) SEGUserInfo *userInfo;
```
Comment (migs647) [inline Analytics/Internal/SEGState.h:34]: ^^

File: Analytics/Internal/SEGState.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+@property (nonatomic, readonly, nonnull) SEGPayloadContext *context;
```
Comment (migs647) [inline Analytics/Internal/SEGState.h:35]: ^^

File: Analytics/Internal/SEGUtils.m [hunk 7]
```diff
@@ -100,3 +112,163 @@ BOOL isUnitTesting()
...
+        dict[@"manufacturer"] = @"Apple";
+        dict[@"type"] = @"ios";
+        dict[@"model"] = getDeviceModel();
+        dict[@"id"] = [[device identifierForVendor] UUIDString];
+        dict[@"name"] = [device model];
+        if (getAdTrackingEnabled(configuration)) {
+            NSString *idfa = configuration.adSupportBlock();
+            // This isn't ideal.  We're doing this because we can't actually check if IDFA is enabled on
+            // the customer device.  Apple docs and tests show that if it is disabled, one gets back all 0's.
+            BOOL adTrackingEnabled = (![idfa isEqualToString:@"00000000-0000-0000-0000-000000000000"]);
```
Comment (migs647) [inline Analytics/Internal/SEGUtils.m:183]: Should we plan for a different format than one with the dashes? Or just react if this ever breaks for some devices / iOS versions?

File: Analytics/Internal/SEGUtils.m [hunk 8]
```diff
@@ -100,3 +112,163 @@ BOOL isUnitTesting()
...
+        @"version" : device.systemVersion
+    };
+
+    CGSize screenSize = [UIScreen mainScreen].bounds.size;
+    dict[@"screen"] = @{
+        @"width" : @(screenSize.width),
+        @"height" : @(screenSize.height)
+    };
+
+// BKS: This bit below doesn't seem to be effective anymore.  Will investigate later.
```
Comment (migs647) [inline Analytics/Internal/SEGUtils.m:207]: 👍🏻

File: AnalyticsTests/MiddlewareTests.swift [hunk 9]
```diff
@@ -138,25 +138,20 @@ class IntegrationMiddlewareTests: QuickSpec {
...
-      }
-      waitUntil(timeout: 60) { done in
-        let queue = DispatchQueue(label: "test")
-        queue.async {
-          while (initialized != true) {
-            sleep(1)
-          }
-          done()
-        }
+      while (!initialized) {
```
Comment (migs647) [inline AnalyticsTests/MiddlewareTests.swift:153]: What thread is this done on?

File: AnalyticsTests/MiddlewareTests.swift [hunk 10]
```diff
@@ -138,25 +138,20 @@ class IntegrationMiddlewareTests: QuickSpec {
...
+        sleep(1)
```
Comment (migs647) [inline AnalyticsTests/MiddlewareTests.swift:154]: Should we add a failure state if this hasn't happened after x amount of time?

File: Analytics/Internal/SEGUtils.m (hunk 7 above)
Comment (bsneed) [reply Analytics/Internal/SEGUtils.m:183]: I'm working off the docs so I think it's ok.  I don't like it all that much, but it makes a good deal of lift on customers if we require them to handle both.

File: AnalyticsTests/MiddlewareTests.swift (hunk 10 above)
Comment (prayansh) [reply AnalyticsTests/MiddlewareTests.swift:154]: ☝️

Comment (bsneed) [reply AnalyticsTests/MiddlewareTests.swift:154]: I think that's something to do once this quick/nimble BS is gone, yes.

**Pull request merged**